File: src/dotnet/Core/Application/Services/VectorMetadataBuilder.cs
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Entities;

namespace Rag.Candidates.Core.Application.Services;

public sealed class VectorMetadataBuilder
{
    private readonly VectorMetadataConfig _config;
    private readonly PrimarySkillsExtractor _skillsExtractor;

    public VectorMetadataBuilder(VectorMetadataConfig config)
    {
        _config = config;
        _skillsExtractor = new PrimarySkillsExtractor(config);
    }

    public Dictionary<string, object> BuildCandidateMetadata(
        CandidateRecord candidate,
        string candidateId,
        string englishLevel,
        int englishLevelNum)
    {
        var metadata = new Dictionary<string, object>
        {
            [_config.FieldType] = _config.TypeCandidate,
            [_config.FieldCandidateId] = candidateId,
            [_config.FieldEnglishLevel] = englishLevel,
            [_config.FieldEnglishLevelNum] = englishLevelNum
        };
        
        // Only set fullname if it exists and is not empty
        // NEVER fallback to candidateId during indexing
        if (!string.IsNullOrWhiteSpace(candidate.GeneralInfo?.Fullname))
        {
            var fullname = candidate.GeneralInfo.Fullname.Trim();
            if (fullname != candidateId)
            {
                metadata[_config.FieldFullname] = fullname;
            }
            else
            {
                Console.WriteLine($"[WARNING] Fullname equals candidateId for {candidateId}, NOT storing in metadata");
            }
        }

        if (candidate.GeneralInfo != null)
        {
            if (candidate.GeneralInfo.SeniorityLevel != null)
                metadata[_config.FieldSeniorityLevel] = candidate.GeneralInfo.SeniorityLevel;

            if (candidate.GeneralInfo.YearsExperience.HasValue)
                metadata[_config.FieldYearsExperience] = candidate.GeneralInfo.YearsExperience.Value;

            if (candidate.GeneralInfo.RelevantYears.HasValue)
                metadata[_config.FieldRelevantYears] = candidate.GeneralInfo.RelevantYears.Value;

            if (candidate.GeneralInfo.MainIndustry != null)
                metadata[_config.FieldMainIndustry] = candidate.GeneralInfo.MainIndustry;
        }

        var primarySkills = _skillsExtractor.Extract(candidate);
        if (primarySkills.Length > 0)
            metadata[_config.FieldPrimarySkills] = primarySkills;

        return metadata;
    }
}

File: src/python/core/application/services/vector_metadata_builder.py
from typing import Dict, Any, Optional
from ...domain.entities.candidate_record import CandidateRecord
from ...domain.configuration.vector_metadata_config import VectorMetadataConfig
from .primary_skills_extractor import PrimarySkillsExtractor


class VectorMetadataBuilder:
    def __init__(self, config: VectorMetadataConfig):
        self.config = config
        self.skills_extractor = PrimarySkillsExtractor(config)
    
    def build_candidate_metadata(
        self,
        candidate: CandidateRecord,
        candidate_id: str,
        english_level: str,
        english_level_num: int
    ) -> Dict[str, Any]:
        metadata = {
            self.config.FIELD_TYPE: self.config.TYPE_CANDIDATE,
            self.config.FIELD_CANDIDATE_ID: candidate_id,
            self.config.FIELD_ENGLISH_LEVEL: english_level,
            self.config.FIELD_ENGLISH_LEVEL_NUM: english_level_num,
        }
        
        if candidate.GeneralInfo:
            metadata[self.config.FIELD_SENIORITY_LEVEL] = (
                candidate.GeneralInfo.SeniorityLevel.value if candidate.GeneralInfo.SeniorityLevel else None
            )
            metadata[self.config.FIELD_YEARS_EXPERIENCE] = (
                candidate.GeneralInfo.YearsExperience if candidate.GeneralInfo.YearsExperience is not None else None
            )
            metadata[self.config.FIELD_RELEVANT_YEARS] = (
                candidate.GeneralInfo.RelevantYears if candidate.GeneralInfo.RelevantYears is not None else None
            )
            metadata[self.config.FIELD_MAIN_INDUSTRY] = (
                candidate.GeneralInfo.MainIndustry or None
            )
        
        # Extract primary skills and serialize as comma-separated string for Chroma
        # (Chroma doesn't support list metadata, unlike Qdrant in .NET)
        primary_skills = self.skills_extractor.extract(candidate)
        metadata[self.config.FIELD_PRIMARY_SKILLS] = ",".join(primary_skills) if primary_skills else ""
        
        return metadata
